File: fenwickTree/gym.py
from sys import stdin, stdout

LSOne = lambda b: b&(-b)


class FenwickTree:

    def __init__ (self, n):
        self.ft = [0]*(n+1)

    def rsq_util(self, b): # return RSQ(1,b)
        s = 0
        while b > 0:
            s += self.ft[b]
            b -= LSOne(b)
        return s

    def rsq(self, a, b):
        return self.rsq_util(b) - (0 if a == 1 else self.rsq_util(a-1))

    def adjust(self, k, v):
        
        while k < len(self.ft):
            self.ft[k] += v
            k += LSOne(v)

if __name__ == "__main__":
    while True:
        # ip, m = list(map(int, stdin.readline().strip().split()))
        ip, m = stdin.readline().split()
        ip = int(ip)
        m = int(m)
        if ip == '':
            break
        count = 0
        l_pc = []
        l_na = []
        for i in range(m):
            # pc, na = list(map(int, stdin.readline().strip().split()))
            pc, na = stdin.readline().split()
            pc = int(pc)
            na = int(na)
            l_pc.append(pc)
            l_na.append(na)

        size = max(l_pc)

        fenwick_tree = FenwickTree(size)
        for pc, na in zip(l_pc, l_na):
            pcmax = size if pc+ip > size else pc+ip
            # pcmin é 1 caso pc-ip <=0, visto que 1 é o menor valor de PC.
            pcmin = 1 if pc-ip <= 0 else pc-ip
            if fenwick_tree.rsq(pcmin, pcmax) <= na:
                count += 1
                fenwick_tree.adjust(pc, 1)
            # Brute force    
            # if len(l):
            #     length = 0
            #     for j in l:
            #         if pc-ip < j <= pc+ip:
            #             length += 1
            #     if length > 0 and length <= na:
            #         l.append(pc)
            # else:
            #     l.append(pc)
        stdout.write(str(count+1)+'\n')

File: fenwickTree/gym.cpp
#include<iostream>
#include<vector>
#include <algorithm>

#define endl '\n'
#define D(x) cout << #x << " = " << (x) << endl;

using namespace std;

typedef vector<int> vi;


class FenwickTree {
    private: 
        vi ft;

    public:

        FenwickTree(int n){
            ft.assign(n+1, 0);
        }

        int rsqUtil(int b){
            int sum = 0;

            for(; b; b -= b&(-b)){
                sum += ft[b];
            }
            return sum;
        }

        int rsq(int a, int b){
            return rsqUtil(b) - (a == 1 ? 0: rsqUtil(a-1));
        }

        void adjust(int k, int v){
            for(; k; k += k&(-k)) ft[k] += v;
        }
};

int main(){
    int ip, m, pc, na, count;

    while(cin >> ip >> m){
        vi l_pc, l_na;

        for (int i = 0; i < m; i++)
        {
            cin >> pc >> na;
            l_pc.push_back(pc);
            l_na.push_back(na);
        }
        int size = *max_element(l_pc.begin(), l_pc.end());

        FenwickTree fenwickTree(size);
        count = 0;
        for (int i = 0; i < size; i++)
        {
            pc = l_pc[i];
            na = l_na[i];

            int pcmax = (pc+ip > size ? size:pc+ip);
            int pcmin = (pc-ip <= 0 ? 1: pc-ip);

            if(fenwickTree.rsq(pcmin, pcmax) <= na) {
                count += 1;
                fenwickTree.adjust(pc, 1);
            }
        }

        cout << count << endl;
        
    }

    return 0;
}